# Goal of the ETL Lesson

The main goal of this ETL pipelines lesson is to take the [World Bank Project data set](https://datacatalog.worldbank.org/dataset/world-bank-projects-operations) and merge this data with the [World Bank indicator data](https://data.worldbank.org/indicator/SP.POP.TOTL). Then you'll load the merged data into a database.

In the process, you'll need to transform these data sets in different ways. And finally, you'll code an ETL pipeline to extract, transform, and load the data all in one step.

# Extracting data from a csv file

The first step in an ETL pipeline is extraction. Data comes in all types of different formats, and you'll practice extracting data from csv files, JSON files, XML files, SQL databases, and the web.

In this first exercise, you'll practice extracting data from a CSV file and then navigating through the results. You'll see that extracting data is not always a straight-forward process.

This exercise contains a series of coding questions for you to solve. If you get stuck, there is a solution file called 1_csv_exercise_solution.ipynb. You can find this solution file by going to File->Open and then clicking on the file name.

# Part 1 projects_data.csv

You'll be using the following csv files:
* projects_data.csv
* population_data.csv

As a first step, try importing the projects data using the pandas [read_csv method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). The file path is just '../data/projects_data.csv'. You can see the file if you click on File->Open in the workspace and open the data folder.

In [70]:
# TODO: import the projects_data.csv file using the pandas library
# Store the results in the df_projects variable
import pandas as pd

In [71]:
df_projects =pd.read_csv('projects_data.csv')

/opt/conda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (44) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Did you get a DType warning? Read about what this warning is in the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.errors.DtypeWarning.html).

Pandas tries to figure out programatically the data type of each column (integer, float, boolean, string). In this case, pandas could not automatically figure out the data type. That is because some columns have more than one possible data types. In other words, this data is messy.

You can use the dtype option to specify the data type of each column. Because there are so many columns in this data set, you can set all columns to be strings at least for now.

Try reading in the data set again using the read_csv() method. This time, also use the option dtype=str so that pandas treats everything like a string.

In [72]:
# TODO: Read in the projects_data.csv file using the read_csv method 
# and dtype = str option
df_projects =pd.read_csv('projects_data.csv',dtype=str)

In [73]:
# Run the cell below to see what the data looks like
df_projects.head()

id  regionname                                        countryname  \
0  P162228       Other                                        World;World   
1  P163962      Africa  Democratic Republic of the Congo;Democratic Re...   
2  P167672  South Asia  People's Republic of Bangladesh;People's Repub...   
3  P158768  South Asia  Islamic  Republic of Afghanistan;Islamic  Repu...   
4  P161364      Africa  Federal Republic of Nigeria;Federal Republic o...   

  prodline                  lendinginstr lendinginstrtype  \
0       RE  Investment Project Financing               IN   
1       PE  Investment Project Financing               IN   
2       PE  Investment Project Financing               IN   
3       PE  Investment Project Financing               IN   
4       PE  Investment Project Financing               IN   

  envassesmentcategorycode supplementprojectflg productlinetype  \
0                        C                    N               L   
1                        B                    N               L   
2                      NaN                    Y               L   
3                        A                    N               L   
4                        B                    N               L   

  projectstatusdisplay     ...     mjtheme3name mjtheme4name mjtheme5name  \
0               Active     ...              NaN          NaN          NaN   
1               Active     ...              NaN          NaN          NaN   
2               Active     ...              NaN          NaN          NaN   
3               Active     ...              NaN          NaN          NaN   
4               Active     ...              NaN          NaN          NaN   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546!$!Ogun State!$!7!$!3.58333!$!NG;000...   

                                            GeoLocID  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546;0002328925;0002565340;0002565343;00...   

                                          GeoLocName                 Latitude  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4  Ogun State;Niger State;Abia State;Edo;Kebbi St...  7;10;5.41667;6.5;11.5;8   

                Longitude            Country Unnamed: 56  
0                     NaN                NaN         NaN  
1                     NaN                NaN         NaN  
2                     NaN                NaN         NaN  
3                     NaN                NaN         NaN  
4  3.58333;6;7.5;6;4;10.5  NG;NG;NG;NG;NG;NG         NaN  

[5 rows x 57 columns]

In [74]:
import numpy as np

In [75]:
df_projects.shape

(18248, 57)

In [76]:
df_projects.isnull().sum()

id                              0
regionname                      0
countryname                     0
prodline                        0
lendinginstr                  246
lendinginstrtype              246
envassesmentcategorycode     5811
supplementprojectflg           53
productlinetype                 0
projectstatusdisplay            4
status                          4
project_name                    0
boardapprovaldate            1504
board_approval_month         1504
closingdate                  3349
lendprojectcost               125
ibrdcommamt                     0
idacommamt                      0
totalamt                        0
grantamt                        0
borrower                     5919
impagency                    6097
url                             0
projectdoc                  18248
majorsector_percent         18248
sector1                         0
sector2                      8721
sector3                     11761
sector4                     13872
sector5       

In [77]:
# TODO: count the number of null values in the data set
# HINT: use the isnull() and sum() methods
np.sum(df_projects.isnull().sum())

479951

Notice that the number 18248 shows up multiple times. There is also a countryname column with 0 missing values and a Country column with 14045 missing values. This data set clearly has some issues that will need to be solved in the transform part of the pipeline.

Next, output the shape of the data frame.

In [78]:
# TODO: output the shape of the data frame
df_projects.shape

(18248, 57)

There are 18248 rows in this data set. Considering many columns had 18248 NaN values, many columns in the data set are filled completely with NaN values.

# Part 2 population_data.csv

Next, use the pandas read_csv method to read in the population_data.csv file. The path to this file is "../data/population_data.csv". When you try to read in this data set using pandas, you'll get an error because there is something wrong with the data.

In [79]:
# TODO: read in the population_data.csv file using the read_csv() method
# Put the results in a variable called df_population
df_population =pd.read_csv('population_data.csv')

ParserError: Error tokenizing data. C error: Expected 3 fields in line 5, saw 63


There is something wrong with this data set. You should see an error that says "expected 3 fields in line 5, saw 63". What might have happened? Try printing out the first few lines of the data file to see what the issue might be.

In [80]:
# TODO: Print out the first 10 lines of the data set, line by line.
# HINT: You can't use the read_csv method from pandas
# HINT: to do this manually, you'll need to use pure Python
# HINT: the open(), readline(), and close() methods should be helpful
# HINT: Use a for loop

filename=open('population_data.csv','r')



In [81]:
File=filename.readlines()

In [82]:
File[:4]

['\ufeff"Data Source","World Development Indicators",\n',
 '\n',
 '"Last Updated Date","2018-06-28",\n',
 '\n']

The first four lines in the file are not properly formatted and don't contain data. Next, read in the data using the read_csv method. But this time, use the skiprows option.

In [83]:
# TODO: read in population data skipping the first four rows
# Put the results in a variable called df_population
df_population =pd.read_csv('population_data.csv',skiprows=4)

In [84]:
# Run this cell to see what the data looks like
df_population.head()

Country Name Country Code     Indicator Name Indicator Code       1960  \
0        Aruba          ABW  Population, total    SP.POP.TOTL    54211.0   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  8996351.0   
2       Angola          AGO  Population, total    SP.POP.TOTL  5643182.0   
3      Albania          ALB  Population, total    SP.POP.TOTL  1608800.0   
4      Andorra          AND  Population, total    SP.POP.TOTL    13411.0   

        1961       1962       1963       1964       1965     ...       \
0    55438.0    56225.0    56695.0    57032.0    57360.0     ...        
1  9166764.0  9345868.0  9533954.0  9731361.0  9938414.0     ...        
2  5753024.0  5866061.0  5980417.0  6093321.0  6203299.0     ...        
3  1659800.0  1711319.0  1762621.0  1814135.0  1864791.0     ...        
4    14375.0    15370.0    16412.0    17469.0    18549.0     ...        

         2009        2010        2011        2012        2013        2014  \
0    101453.0    101669.0    102053.0    102577.0    103187.0    103795.0   
1  28004331.0  28803167.0  29708599.0  30696958.0  31731688.0  32758020.0   
2  22549547.0  23369131.0  24218565.0  25096150.0  25998340.0  26920466.0   
3   2927519.0   2913021.0   2905195.0   2900401.0   2895092.0   2889104.0   
4     84462.0     84449.0     83751.0     82431.0     80788.0     79223.0   

         2015        2016        2017  Unnamed: 62  
0    104341.0    104822.0    105264.0          NaN  
1  33736494.0  34656032.0  35530081.0          NaN  
2  27859305.0  28813463.0  29784193.0          NaN  
3   2880703.0   2876101.0   2873457.0          NaN  
4     78014.0     77281.0     76965.0          NaN  

[5 rows x 63 columns]

Make sure to scroll over to see what the last column looks like. That last column, called 'Unnamed: 62', doesn't look very useful and is filled with NaN values.

In [85]:
# TODO: Count the number of null values in each column
df_population.isnull().sum()

Country Name        0
Country Code        0
Indicator Name      0
Indicator Code      0
1960                4
1961                4
1962                4
1963                4
1964                4
1965                4
1966                4
1967                4
1968                4
1969                4
1970                4
1971                4
1972                4
1973                4
1974                4
1975                4
1976                4
1977                4
1978                4
1979                4
1980                4
1981                4
1982                4
1983                4
1984                4
1985                4
                 ... 
1989                4
1990                2
1991                2
1992                3
1993                3
1994                3
1995                2
1996                2
1997                2
1998                1
1999                1
2000                1
2001                1
2002                1
2003      

It looks like every year column has at least one NaN value. 

In [86]:
# TODO: Count the number of null values in each row
# HINT: In the sum method, use axis=1
df_projects.isnull().sum(axis=1)

0        35
1        34
2        36
3        32
4        23
5        33
6        34
7        33
8        34
9        33
10       34
11       36
12       35
13       36
14       35
15       33
16       36
17       34
18       33
19       33
20       35
21       33
22       33
23       36
24       35
25       30
26       30
27       35
28       33
29       29
         ..
18218    33
18219    33
18220    33
18221    33
18222    33
18223    35
18224    33
18225    33
18226    33
18227    33
18228    33
18229    33
18230    33
18231    33
18232    33
18233    33
18234    33
18235    33
18236    33
18237    33
18238    33
18239    33
18240    33
18241    33
18242    33
18243    33
18244    33
18245    33
18246    33
18247    33
Length: 18248, dtype: int64

And it looks like almost every row has only one null value. That is probably from the 'Unnamed: 62' column that doesn't have any relevant information in it. Next, drop the 'Unnamed: 62' column from the data frame.

In [88]:
# TODO: drop the 'Unnamed: 62' column from the data frame, 
# and save the results in the df_population variable


df_population=df_population.drop(columns=['Unnamed: 62'])

In [89]:
df_population

Country Name Country Code  \
0                                                Aruba          ABW   
1                                          Afghanistan          AFG   
2                                               Angola          AGO   
3                                              Albania          ALB   
4                                              Andorra          AND   
5                                           Arab World          ARB   
6                                 United Arab Emirates          ARE   
7                                            Argentina          ARG   
8                                              Armenia          ARM   
9                                       American Samoa          ASM   
10                                 Antigua and Barbuda          ATG   
11                                           Australia          AUS   
12                                             Austria          AUT   
13                                          Azerbaijan          AZE   
14                                             Burundi          BDI   
15                                             Belgium          BEL   
16                                               Benin          BEN   
17                                        Burkina Faso          BFA   
18                                          Bangladesh          BGD   
19                                            Bulgaria          BGR   
20                                             Bahrain          BHR   
21                                        Bahamas, The          BHS   
22                              Bosnia and Herzegovina          BIH   
23                                             Belarus          BLR   
24                                              Belize          BLZ   
25                                             Bermuda          BMU   
26                                             Bolivia          BOL   
27                                              Brazil          BRA   
28                                            Barbados          BRB   
29                                   Brunei Darussalam          BRN   
..                                                 ...          ...   
234  Latin America & the Caribbean (IDA & IBRD coun...          TLA   
235                                        Timor-Leste          TLS   
236  Middle East & North Africa (IDA & IBRD countries)          TMN   
237                                              Tonga          TON   
238                            South Asia (IDA & IBRD)          TSA   
239          Sub-Saharan Africa (IDA & IBRD countries)          TSS   
240                                Trinidad and Tobago          TTO   
241                                            Tunisia          TUN   
242                                             Turkey          TUR   
243                                             Tuvalu          TUV   
244                                           Tanzania          TZA   
245                                             Uganda          UGA   
246                                            Ukraine          UKR   
247                                Upper middle income          UMC   
248                                            Uruguay          URY   
249                                      United States          USA   
250                                         Uzbekistan          UZB   
251                     St. Vincent and the Grenadines          VCT   
252                                      Venezuela, RB          VEN   
253                             British Virgin Islands          VGB   
254                              Virgin Islands (U.S.)          VIR   
255                                            Vietnam          VNM   
256                                            Vanuatu          VUT   
257                                              World          WLD   
258                                              Samoa          WS

In [91]:
# Run this code cell.  
# This code outputs any row that contains a null value
# The purpose is to see what rows contain null values now that 
#   'Unnamed: 62' was dropped from the data.
df_population[df_population.isnull().any(axis=1)]

Country Name Country Code     Indicator Name Indicator Code  \
67                     Eritrea          ERI  Population, total    SP.POP.TOTL   
108             Not classified          INX  Population, total    SP.POP.TOTL   
125                     Kuwait          KWT  Population, total    SP.POP.TOTL   
194         West Bank and Gaza          PSE  Population, total    SP.POP.TOTL   
212                     Serbia          SRB  Population, total    SP.POP.TOTL   
223  Sint Maarten (Dutch part)          SXM  Population, total    SP.POP.TOTL   

          1960       1961       1962       1963       1964       1965  \
67   1397491.0  1432640.0  1469645.0  1508273.0  1548187.0  1589179.0   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125   269618.0   301336.0   338296.0   379891.0   425235.0   473554.0   
194        NaN        NaN        NaN        NaN        NaN        NaN   
212        NaN        NaN        NaN        NaN        NaN        NaN   
223        NaN        NaN        NaN        NaN        NaN        NaN   

       ...           2008       2009       2010       2011       2012  \
67     ...      4232636.0  4310334.0  4390840.0  4474690.0        NaN   
108    ...            NaN        NaN        NaN        NaN        NaN   
125    ...      2652340.0  2818939.0  2998083.0  3191051.0  3395556.0   
194    ...      3596688.0  3702218.0  3811102.0  3927051.0  4046901.0   
212    ...      7350222.0  7320807.0  7291436.0  7234099.0  7199077.0   
223    ...        33964.0    34238.0    34056.0    33435.0    34640.0   

          2013       2014       2015       2016       2017  
67         NaN        NaN        NaN        NaN        NaN  
108        NaN        NaN        NaN        NaN        NaN  
125  3598385.0  3782450.0  3935794.0  4052584.0  4136528.0  
194  4169506.0  4294682.0  4422143.0  4551566.0  4684777.0  
212  7164132.0  7130576.0  7095383.0  7058322.0  7022268.0  
223    36607.0    37685.0    38824.0    39969.0    41109.0  

[6 rows x 62 columns]

# Conclusion 

This population data doesn't look too bad. Only six rows have missing values. In the transformation part of the lesson, you'll have to deal with these missing values somehow.

If you would like to see the solution file for this exercise, go to File->Open and click on 1_csv_exercise_solution.ipynb.

In the next exercise, you'll practice extracting data json and xml files.